### ISSUE TITLE
type_conversion::to_base() is called even if i_null is explicitly passed to use()

### ISSUE THREAD

**Issue by ddurham2** [SOCI/soci#802]
Given the following statement from the documentation (http://soci.sourceforge.net/doc/release/4.0/indicators/#insert-with-null-values)

> Indicator variables can be also used when binding input data, to control whether the data is to be used as provided, **or explicitly overrided to be null**

e.g. 
```
    auto ind = soci::i_null;
    use(x, ind);
```

If you are using a custom type_conversion<>, to_base() is called even if the explicitly given indicator is i_null.  This seems to contradict the docs.

We have a type_conversion<> for boost::any. In to_base(), we make an empty boost::any return i_null for the indicator.  This is the same as your implementation for boost::optional<>.

But since to_base() is called regardless of the value of an indicator given to use(), the boost::any's value gets put into the database, overriding that i_null was passed to use().   I verified that boost::optional<> behaves the same way, which is erroneous, IMO.

I would propose that a condition be added to conversion_use_type<>::convert_to_base like so:
```
   void convert_to_base() {
        if (&ind_ != &ownInd_ && ind_ == soci::i_null) {
             // An indicator was explicitly passed to soci::use(...), and it species i_null.
             // Therefore, NULL should be put into the DB, and there is no reason to convert the value.
             return;
        }

        type_conversion<T>::to_base(value_,
            base_value_holder<T>::val_, ind_);
    }
```
Does that seem reasonable? Or am I missing something?

NOTE: I'm actually using 3.2 but looking to upgrade to 4.0 to fix such issues, but didn't see that the behavior had changed.

**Comment by vadz** [comment]
I agree that I'd expect things to work like this and I'm also surprised that they don't. And the fix also seems reasonable, although I don't really understand why do we need to check for not using `ownInd_` explicitly here as it seems to me that it would be never set to `i_null` in any case, but, of course, this is harmless  in any case and in the worst case just useless.

If you can take the time to add a new test (see `tests/common-tests.h`) checking that `use(foo, i_null)` works as expected and make a PR with this change, I'd be glad to merge it. TIA!

**Comment by ddurham2** [comment]
What that 'if' is trying to say is that if an indicator was passed in to use (&ind_ != &ownInd_) AND what was passed in is presently set to i_null (ind_ == i_null), then that is when we don't need to call to_base().

I guess you're saying we don't need the != because it would never be the case that the ownInd_ would be null..  That's true.  I was just trying to be explicit.  Perhaps in 3.2 (where I'm making this fix) I was concerned about the initialized value of ownInd_.  There were some problems earler on where it wasn't initialized consistently.  I think I like you're way better tho.

**Comment by vadz** [comment]
> I guess you're saying we don't need the != because it would never be the case that the ownInd_ would be null..

Yes, exactly, sorry for being unclear.

**Comment by ddurham2** [comment]
Actually, on further analysis, I believe (and I've seen in 3.2.2) that a prepared statement can lead to ownInd_'s value possibly becoming i_null.  

With prepared statements, soci invokes type_conversions each time they are executed.

So, if a prepared statement uses `use(x)` to bind to variable x, then x's value is converted each time the statement is executed, and in this case, a reference to ownInd_ would be passed to to_base() and it's value could be altered each time.  

If an indicator is supplied by the user (i.e. &ind_ != &ownInd_) , then we trust _it_ to indicate null-ness (overriding anything else.. as the docs state).   Else, if it wasn't supplied by the user, then we need to invoke to_base() each time, letting null-ness be determined by the value itself (which in the case of simple types cannot be null.. so I trust that their to_base() impls will always set ind _to i_ok or else internally short-circuit that consideration elsewhere).

Hence, I believe it's necessary to keep the != part.  Make sense?

**Comment by ddurham2** [comment]
@vadz, does that (my prev comment) stand to reason?

**Comment by vadz** [comment]
Sorry, I just didn't have time to return to this... will try to do a.s.a.p.

**Comment by vadz** [comment]
I've finally reread the code once again and I'm still not sure a situation in which `ownInd_` could end up being set to `i_null` could happen: as we're only speaking about `use()` here, and not `into()`, it seems to me that `ownInd_` should keep its initial `i_ok` value no matter what.

But, frankly, this is not that important, I could be missing something and we could really need this `!= ownInd_` comparison and it's not a problem to keep it. The really important things are:

1. Add the currently failing unit test.
1. Change the code as you propose and check that it fixes the test added above.
1. Run the full unit test suite to check that it doesn't break something else.
1. Make a PR with the changes so that it could be merged.

If you could please do this, it'd be great. TIA!

**Comment by ddurham2** [comment]
I will attempt a PR and demonstration.

But, it should be apparent that it could happen when a prepared statement (executed multiple times) is combined with a custom conversion (which is possibly getting passed _ownInd when use() isn't passed one).  Its value can get changed by the custom conversion (perhaps not every time) and won't necessarily stay its initial value (each time the prepared statement is executed).

**Comment by vadz** [comment]
Yes, you're right, sorry, I've somehow overlooked this.

Just to be clear, there is no need to demonstrate this, by "the currently failing test" I just meant a test doing what you wrote in the initial comment, i.e.

    auto ind = soci::i_null;
    use(x, ind);

and checking that custom `to_base()` is not called in this case.
